File: scripts/agl_health_client.py
# SPDX-FileCopyrightText: (c) 2026 Toyota Connected North America
# SPDX-License-Identifier: GPLv3
#!/usr/bin/env python3
"""Exercise the AglHealth capnp plugin against a running TestRunnerService."""

import argparse
import asyncio
import os

import capnp

capnp.remove_import_hook()
script_dir = os.path.dirname(os.path.abspath(__file__))
repo_root = os.path.normpath(os.path.join(script_dir, '..'))
# Load agl_health.capnp directly — it has no imports so capnp's filesystem
# never encounters ".." path components (which KJ refuses to resolve).
# The server's bootstrap capability implements AglHealth transitively through
# TestRunnerService -> Plugins -> AglHealth, so the cast works fine.
schema_path = os.path.join(repo_root, 'plugins', 'agl_health', 'agl_health.capnp')
agl_health_capnp = capnp.load(schema_path)


def fmt_bytes(n):
    for unit in ('B', 'KiB', 'MiB', 'GiB'):
        if n < 1024:
            return f"{n:.1f} {unit}"
        n /= 1024
    return f"{n:.1f} TiB"


def fmt_ns(n):
    if n < 1_000:
        return f"{n} ns"
    if n < 1_000_000:
        return f"{n/1_000:.2f} µs"
    if n < 1_000_000_000:
        return f"{n/1_000_000:.2f} ms"
    return f"{n/1_000_000_000:.3f} s"


def print_memory(m):
    print("=== Memory ===")
    print(f"  Total:         {fmt_bytes(m.totalBytes)}")
    print(f"  Free:          {fmt_bytes(m.freeBytes)}")
    print(f"  Cached:        {fmt_bytes(m.cachedBytes)}")
    print(f"  Buffered:      {fmt_bytes(m.bufferedBytes)}")
    print(f"  Slab:          {fmt_bytes(m.slabBytes)}")
    print(f"  Swap used:     {fmt_bytes(m.swapUsedBytes)}")
    print(f"  Swap free:     {fmt_bytes(m.swapFreeBytes)}")
    print(f"  Page faults (minor/major): {m.pageFaultsMinor} / {m.pageFaultsMajor}")
    print(f"  PSI some:      {m.psiSomePctX100 / 100:.2f}%")
    print(f"  PSI full:      {m.psiFullPctX100 / 100:.2f}%")
    print(f"  OOM kills:     {m.oomKillsTotal}")


def print_cpu(c):
    print("=== CPU ===")
    print(f"  Load avg:  {c.load1:.2f}  {c.load5:.2f}  {c.load15:.2f}  (1/5/15 min)")
    for core in c.cores:
        print(f"  CPU{core.cpuId}: user={fmt_ns(core.userNs)} sys={fmt_ns(core.systemNs)}"
              f" iowait={fmt_ns(core.iowaitNs)} idle={fmt_ns(core.idleNs)}"
              f" ctx={core.ctxSwitches}")


def print_processes(procs):
    print("=== Processes ===")
    print(f"  {'PID':>7}  {'PPID':>7}  {'UID':>5}  {'THR':>4}  {'RSS':>10}  {'FDs':>4}  COMM")
    for p in procs:
        print(f"  {p.pid:>7}  {p.ppid:>7}  {p.uid:>5}  {p.threadCount:>4}"
              f"  {fmt_bytes(p.memRssBytes):>10}  {p.openFds:>4}  {p.comm}")


def print_network(n):
    print("=== Network ===")
    print("  Interfaces:")
    for iface in n.ifaces:
        print(f"    [{iface.ifaceIdx}] {iface.name}: "
              f"rx={fmt_bytes(iface.rxBytes)} ({iface.rxPackets} pkts, {iface.rxErrors} err, {iface.rxDrops} drop) "
              f"tx={fmt_bytes(iface.txBytes)} ({iface.txPackets} pkts, {iface.txErrors} err, {iface.txDrops} drop)")
    t = n.tcp
    print("  TCP states:")
    print(f"    ESTABLISHED={t.established}  SYN_SENT={t.synSent}  SYN_RECV={t.synRecv}")
    print(f"    FIN_WAIT1={t.finWait1}  FIN_WAIT2={t.finWait2}  TIME_WAIT={t.timeWait}  CLOSE_WAIT={t.closeWait}")
    print(f"    LISTEN={t.listen}  listen_overflows={t.listenOverflows}")
    print(f"    retransmits={t.retransmits}  resets_in={t.resetsIn}  resets_out={t.resetsOut}")


def print_security(s):
    print("=== Security ===")
    print(f"  ptrace:         {s.ptrace}")
    print(f"  memfd_create:   {s.memfdCreate}")
    print(f"  prctl:          {s.prctl}")
    print(f"  setuid:         {s.setuid}")
    print(f"  exec_anomaly:   {s.execAnomaly}")
    print(f"  capability_use: {s.capabilityUse}")


def print_scheduler(s):
    print("=== Scheduler ===")
    h = s.histogram
    print(f"  Latency histogram:")
    print(f"    <10µs={h.lt10Us}  <100µs={h.lt100Us}  <1ms={h.lt1Ms}  <10ms={h.lt10Ms}")
    print(f"    <100ms={h.lt100Ms}  <1s={h.lt1S}  <10s={h.lt10S}  >=10s={h.ge10S}")
    print(f"  Total events: {s.totalCount}")
    if s.totalCount:
        print(f"  Avg latency:  {fmt_ns(s.totalLatNs // s.totalCount)}")
    print(f"  Max latency:  {fmt_ns(s.maxLatNs)}")
    print(f"  p50={fmt_ns(s.p50Ns)}  p95={fmt_ns(s.p95Ns)}  p99={fmt_ns(s.p99Ns)}")


COMMANDS = {
    'metrics', 'memory', 'cpu', 'processes', 'network', 'security', 'scheduler',
}


def parse_args():
    parser = argparse.ArgumentParser(
        description="AglHealth capnp plugin client",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=f"Commands: {', '.join(sorted(COMMANDS))}  (default: metrics)",
    )
    parser.add_argument("--host", default='localhost:4004', help="HOST:PORT (default: localhost:4004)")
    parser.add_argument("--daemon-url", metavar="URL",
                        help="Override the agl-health daemon URL on the server before querying")
    parser.add_argument("--limit", type=int, default=100,
                        help="Max processes to return for 'processes' command (default: 100)")
    parser.add_argument("command", nargs='?', default='metrics',
                        choices=sorted(COMMANDS),
                        help="Which subsystem to query")
    return parser.parse_args()


async def run(connection, args):
    client = capnp.TwoPartyClient(connection)
    server = client.bootstrap().cast_as(agl_health_capnp.AglHealth)

    if args.daemon_url:
        await server.setDaemonUrl(url=args.daemon_url)
        print(f"Daemon URL set to: {args.daemon_url}")

    cmd = args.command
    if cmd == 'metrics':
        resp = await server.getMetrics()
        snap = resp.snapshot
        print(f"Timestamp: {snap.timestampNs} ns")
        print_memory(snap.memory)
        print_cpu(snap.cpu)
        print_processes(snap.processes)
        print_network(snap.network)
        print_security(snap.security)
        print_scheduler(snap.scheduler)
    elif cmd == 'memory':
        resp = await server.getMemory()
        print_memory(resp.memory)
    elif cmd == 'cpu':
        resp = await server.getCpu()
        print_cpu(resp.cpu)
    elif cmd == 'processes':
        resp = await server.getProcesses(limit=args.limit)
        print_processes(resp.processes)
    elif cmd == 'network':
        resp = await server.getNetwork()
        print_network(resp.network)
    elif cmd == 'security':
        resp = await server.getSecurity()
        print_security(resp.security)
    elif cmd == 'scheduler':
        resp = await server.getScheduler()
        print_scheduler(resp.scheduler)


async def cmd_main(args):
    host, port = args.host.rsplit(':', 1)
    conn = await capnp.AsyncIoStream.create_connection(host=host, port=int(port))
    await run(conn, args)


if __name__ == "__main__":
    args = parse_args()
    asyncio.run(capnp.run(cmd_main(args)))

File: plugins/agl_health/AglHealth_Client.cpp
// SPDX-FileCopyrightText: (c) 2026 Toyota Connected North America
// SPDX-License-Identifier: GPLv3
#include "AglHealth_Client.h"
#include <kj/debug.h>
#include "spdlog/spdlog.h"

// ---------------------------------------------------------------------------
// Decode helpers — capnp reader → plain C++ struct
// ---------------------------------------------------------------------------

static AglMemorySnapshot decode_memory(::MemorySnapshot::Reader r) {
  AglMemorySnapshot m;
  m.total_bytes        = r.getTotalBytes();
  m.free_bytes         = r.getFreeBytes();
  m.cached_bytes       = r.getCachedBytes();
  m.buffered_bytes     = r.getBufferedBytes();
  m.slab_bytes         = r.getSlabBytes();
  m.swap_used_bytes    = r.getSwapUsedBytes();
  m.swap_free_bytes    = r.getSwapFreeBytes();
  m.page_faults_minor  = r.getPageFaultsMinor();
  m.page_faults_major  = r.getPageFaultsMajor();
  m.psi_some_pct_x100  = r.getPsiSomePctX100();
  m.psi_full_pct_x100  = r.getPsiFullPctX100();
  m.oom_kills_total    = r.getOomKillsTotal();
  return m;
}

static AglCpuSnapshot decode_cpu(::CpuSnapshot::Reader r) {
  AglCpuSnapshot cpu;
  cpu.load_1  = r.getLoad1();
  cpu.load_5  = r.getLoad5();
  cpu.load_15 = r.getLoad15();
  for (auto core : r.getCores()) {
    AglCoreStats c;
    c.cpu_id       = core.getCpuId();
    c.user_ns      = core.getUserNs();
    c.system_ns    = core.getSystemNs();
    c.iowait_ns    = core.getIowaitNs();
    c.irq_ns       = core.getIrqNs();
    c.softirq_ns   = core.getSoftirqNs();
    c.idle_ns      = core.getIdleNs();
    c.ctx_switches = core.getCtxSwitches();
    cpu.cores.push_back(c);
  }
  return cpu;
}

static std::vector<AglProcessStats> decode_processes(
    capnp::List<::ProcessStats>::Reader list) {
  std::vector<AglProcessStats> out;
  out.reserve(list.size());
  for (auto p : list) {
    AglProcessStats s;
    s.pid               = p.getPid();
    s.ppid              = p.getPpid();
    s.uid               = p.getUid();
    s.thread_count      = p.getThreadCount();
    s.cpu_user_ns       = p.getCpuUserNs();
    s.cpu_system_ns     = p.getCpuSystemNs();
    s.mem_rss_bytes     = p.getMemRssBytes();
    s.mem_vms_bytes     = p.getMemVmsBytes();
    s.voluntary_ctx_sw  = p.getVoluntaryCtxSw();
    s.involuntary_ctx_sw= p.getInvoluntaryCtxSw();
    s.read_bytes        = p.getReadBytes();
    s.write_bytes       = p.getWriteBytes();
    s.page_faults_minor = p.getPageFaultsMinor();
    s.page_faults_major = p.getPageFaultsMajor();
    s.start_time_ns     = p.getStartTimeNs();
    s.open_fds          = p.getOpenFds();
    s.comm              = p.getComm();
    out.push_back(std::move(s));
  }
  return out;
}

static AglNetworkSnapshot decode_network(::NetworkSnapshot::Reader r) {
  AglNetworkSnapshot net;
  for (auto iface : r.getIfaces()) {
    AglNetIfaceStats s;
    s.iface_idx  = iface.getIfaceIdx();
    s.name       = iface.getName();
    s.rx_bytes   = iface.getRxBytes();
    s.tx_bytes   = iface.getTxBytes();
    s.rx_packets = iface.getRxPackets();
    s.tx_packets = iface.getTxPackets();
    s.rx_drops   = iface.getRxDrops();
    s.tx_drops   = iface.getTxDrops();
    s.rx_errors  = iface.getRxErrors();
    s.tx_errors  = iface.getTxErrors();
    net.ifaces.push_back(std::move(s));
  }
  auto tcp = r.getTcp();
  net.tcp.established     = tcp.getEstablished();
  net.tcp.syn_sent        = tcp.getSynSent();
  net.tcp.syn_recv        = tcp.getSynRecv();
  net.tcp.fin_wait1       = tcp.getFinWait1();
  net.tcp.fin_wait2       = tcp.getFinWait2();
  net.tcp.time_wait       = tcp.getTimeWait();
  net.tcp.close_wait      = tcp.getCloseWait();
  net.tcp.listen          = tcp.getListen();
  net.tcp.listen_overflows= tcp.getListenOverflows();
  net.tcp.retransmits     = tcp.getRetransmits();
  net.tcp.resets_in       = tcp.getResetsIn();
  net.tcp.resets_out      = tcp.getResetsOut();
  return net;
}

static AglSecuritySnapshot decode_security(::SecuritySnapshot::Reader r) {
  AglSecuritySnapshot s;
  s.ptrace        = r.getPtrace();
  s.memfd_create  = r.getMemfdCreate();
  s.prctl         = r.getPrctl();
  s.setuid        = r.getSetuid();
  s.exec_anomaly  = r.getExecAnomaly();
  s.capability_use= r.getCapabilityUse();
  return s;
}

static AglSchedulerSnapshot decode_scheduler(::SchedulerSnapshot::Reader r) {
  AglSchedulerSnapshot s;
  auto h = r.getHistogram();
  s.histogram.lt_10us  = h.getLt10Us();
  s.histogram.lt_100us = h.getLt100Us();
  s.histogram.lt_1ms   = h.getLt1Ms();
  s.histogram.lt_10ms  = h.getLt10Ms();
  s.histogram.lt_100ms = h.getLt100Ms();
  s.histogram.lt_1s    = h.getLt1S();
  s.histogram.lt_10s   = h.getLt10S();
  s.histogram.ge_10s   = h.getGe10S();
  s.total_count  = r.getTotalCount();
  s.total_lat_ns = r.getTotalLatNs();
  s.max_lat_ns   = r.getMaxLatNs();
  s.p50_ns       = r.getP50Ns();
  s.p95_ns       = r.getP95Ns();
  s.p99_ns       = r.getP99Ns();
  return s;
}

// ---------------------------------------------------------------------------
// RPC calls
// ---------------------------------------------------------------------------

AglMetricSnapshot PluginAglHealthClient::GetMetrics() {
  KJ_REQUIRE(isConnected, "Not connected to Test Runner server");
  auto client = getMain<AglHealth>();
  auto response = waitWithTimeout(client.getMetricsRequest().send());
  auto snap = response.getSnapshot();
  AglMetricSnapshot out;
  out.timestamp_ns = snap.getTimestampNs();
  out.memory    = decode_memory(snap.getMemory());
  out.cpu       = decode_cpu(snap.getCpu());
  out.processes = decode_processes(snap.getProcesses());
  out.network   = decode_network(snap.getNetwork());
  out.security  = decode_security(snap.getSecurity());
  out.scheduler = decode_scheduler(snap.getScheduler());
  return out;
}

AglMemorySnapshot PluginAglHealthClient::GetMemory() {
  KJ_REQUIRE(isConnected, "Not connected to Test Runner server");
  auto client = getMain<AglHealth>();
  auto response = waitWithTimeout(client.getMemoryRequest().send());
  return decode_memory(response.getMemory());
}

AglCpuSnapshot PluginAglHealthClient::GetCpu() {
  KJ_REQUIRE(isConnected, "Not connected to Test Runner server");
  auto client = getMain<AglHealth>();
  auto response = waitWithTimeout(client.getCpuRequest().send());
  return decode_cpu(response.getCpu());
}

std::vector<AglProcessStats> PluginAglHealthClient::GetProcesses(uint32_t limit) {
  KJ_REQUIRE(isConnected, "Not connected to Test Runner server");
  auto client = getMain<AglHealth>();
  auto req = client.getProcessesRequest();
  req.setLimit(limit);
  auto response = waitWithTimeout(req.send());
  return decode_processes(response.getProcesses());
}

AglNetworkSnapshot PluginAglHealthClient::GetNetwork() {
  KJ_REQUIRE(isConnected, "Not connected to Test Runner server");
  auto client = getMain<AglHealth>();
  auto response = waitWithTimeout(client.getNetworkRequest().send());
  return decode_network(response.getNetwork());
}

AglSecuritySnapshot PluginAglHealthClient::GetSecurity() {
  KJ_REQUIRE(isConnected, "Not connected to Test Runner server");
  auto client = getMain<AglHealth>();
  auto response = waitWithTimeout(client.getSecurityRequest().send());
  return decode_security(response.getSecurity());
}

AglSchedulerSnapshot PluginAglHealthClient::GetScheduler() {
  KJ_REQUIRE(isConnected, "Not connected to Test Runner server");
  auto client = getMain<AglHealth>();
  auto response = waitWithTimeout(client.getSchedulerRequest().send());
  return decode_scheduler(response.getScheduler());
}

void PluginAglHealthClient::SetDaemonUrl(const std::string& url) {
  KJ_REQUIRE(isConnected, "Not connected to Test Runner server");
  auto client = getMain<AglHealth>();
  auto req = client.setDaemonUrlRequest();
  req.setUrl(url);
  waitWithTimeout(req.send());
  spdlog::debug("AglHealth daemon URL set to: {}", url);
}
